File: clib/examples/09_snake.cpp
// 09_snake.cpp - Snake Game
//
// Classic Snake: use arrow keys to control the snake, eat food to grow,
// game over if you hit the wall or yourself.
// Learn: DrawGrid, FillCell, IsKeyPressed, game state machine, timed movement
//
// Win32: g++ -o 09_snake.exe 09_snake.cpp -mwindows
// SDL:   g++ -DUSE_SDL -std=c++11 -O2 -o 09_snake.exe 09_snake.cpp -lSDL2

#if defined(_WIN32) && !defined(USE_SDL)
#include "../GameLib.h"
#else
#include "../GameLib.SDL.h"
#endif

#define GRID_ROWS 20
#define GRID_COLS 20
#define CELL_SIZE 22
#define MAX_SNAKE 400

int main()
{
    GameLib game;

    int gridW = GRID_COLS * CELL_SIZE;
    int gridH = GRID_ROWS * CELL_SIZE;
    int winW = gridW + 160;  // leave space for info panel on right
    int winH = gridH + 40;   // leave space for title on top
    game.Open(winW, winH, "09 - Snake", true);

    int gridX = 10, gridY = 30;

    // Snake body (store row/col for each segment in array)
    int snakeR[MAX_SNAKE], snakeC[MAX_SNAKE];
    int snakeLen = 3;

    // Initial position
    snakeR[0] = 10; snakeC[0] = 10;
    snakeR[1] = 10; snakeC[1] = 9;
    snakeR[2] = 10; snakeC[2] = 8;

    // Direction: 0=up 1=down 2=left 3=right
    int dir = 3;
    int nextDir = 3;

    // Food
    int foodR = 5, foodC = 15;

    int score = 0;
    bool gameOver = false;
    bool paused = false;

    // Movement timer
    double moveTimer = 0.0;
    double moveInterval = 0.15;

    while (!game.IsClosed()) {
        if (game.IsKeyPressed(KEY_ESCAPE)) break;

        double dt = game.GetDeltaTime();

        if (!gameOver) {
            // Pause
            if (game.IsKeyPressed(KEY_P))
                paused = !paused;

            if (!paused) {
                // Direction input (cannot reverse directly)
                if (game.IsKeyPressed(KEY_UP)    && dir != 1) nextDir = 0;
                if (game.IsKeyPressed(KEY_DOWN)  && dir != 0) nextDir = 1;
                if (game.IsKeyPressed(KEY_LEFT)  && dir != 3) nextDir = 2;
                if (game.IsKeyPressed(KEY_RIGHT) && dir != 2) nextDir = 3;

                // Timed movement
                moveTimer += dt;
                if (moveTimer >= moveInterval) {
                    moveTimer = 0;
                    dir = nextDir;

                    // Calculate new head position
                    int newR = snakeR[0];
                    int newC = snakeC[0];
                    if (dir == 0) newR--;
                    if (dir == 1) newR++;
                    if (dir == 2) newC--;
                    if (dir == 3) newC++;

                    // Wall collision
                    if (newR < 0 || newR >= GRID_ROWS || newC < 0 || newC >= GRID_COLS) {
                        gameOver = true;
                    } else {
                        // Self collision
                        for (int i = 0; i < snakeLen; i++) {
                            if (snakeR[i] == newR && snakeC[i] == newC) {
                                gameOver = true;
                                break;
                            }
                        }
                    }

                    if (!gameOver) {
                        // Eat food?
                        bool ate = (newR == foodR && newC == foodC);

                        // Move snake body (copy from tail forward)
                        if (!ate) {
                            for (int i = snakeLen - 1; i > 0; i--) {
                                snakeR[i] = snakeR[i - 1];
                                snakeC[i] = snakeC[i - 1];
                            }
                        } else {
                            // Grow: first shift all segments back, make room for head
                            if (snakeLen < MAX_SNAKE) snakeLen++;
                            for (int i = snakeLen - 1; i > 0; i--) {
                                snakeR[i] = snakeR[i - 1];
                                snakeC[i] = snakeC[i - 1];
                            }
                            score += 10;
                            // Speed up
                            if (moveInterval > 0.06) moveInterval -= 0.003;

                            // New food (ensure not on snake)
                            bool onSnake;
                            do {
                                foodR = GameLib::Random(0, GRID_ROWS - 1);
                                foodC = GameLib::Random(0, GRID_COLS - 1);
                                onSnake = false;
                                for (int i = 0; i < snakeLen; i++) {
                                    if (snakeR[i] == foodR && snakeC[i] == foodC) {
                                        onSnake = true;
                                        break;
                                    }
                                }
                            } while (onSnake);
                        }
                        snakeR[0] = newR;
                        snakeC[0] = newC;
                    }
                }
            }
        } else {
            if (game.IsKeyPressed(KEY_R)) {
                snakeLen = 3;
                snakeR[0] = 10; snakeC[0] = 10;
                snakeR[1] = 10; snakeC[1] = 9;
                snakeR[2] = 10; snakeC[2] = 8;
                dir = 3; nextDir = 3;
                foodR = 5; foodC = 15;
                score = 0;
                moveInterval = 0.15;
                gameOver = false;
            }
        }

        // --- Drawing ---
        game.Clear(COLOR_BLACK);

        // Title
        game.DrawTextScale(gridX, 5, "SNAKE", COLOR_GREEN, 16, 16);

        // Grid
        game.DrawGrid(gridX, gridY, GRID_ROWS, GRID_COLS, CELL_SIZE, COLOR_DARK_GRAY);

        // Food (red)
        game.FillCell(gridX, gridY, foodR, foodC, CELL_SIZE, COLOR_RED);

        // Snake body
        for (int i = 0; i < snakeLen; i++) {
            uint32_t c = (i == 0) ? COLOR_GREEN : COLOR_DARK_GREEN;
            game.FillCell(gridX, gridY, snakeR[i], snakeC[i], CELL_SIZE, c);
        }

        // Right info panel
        int infoX = gridX + gridW + 15;
        game.DrawText(infoX, 40, "Score:", COLOR_WHITE);
        game.DrawNumber(infoX, 55, score, COLOR_GOLD);

        game.DrawText(infoX, 85, "Length:", COLOR_WHITE);
        game.DrawNumber(infoX, 100, snakeLen, COLOR_CYAN);

        game.DrawText(infoX, 140, "Controls:", COLOR_GRAY);
        game.DrawText(infoX, 158, "Arrows", COLOR_LIGHT_GRAY);
        game.DrawText(infoX, 172, "P: Pause", COLOR_LIGHT_GRAY);

        if (paused && !gameOver) {
            game.FillRect(gridX + gridW / 2 - 50, gridY + gridH / 2 - 15, 100, 30, COLOR_DARK_GRAY);
            game.DrawText(gridX + gridW / 2 - 30, gridY + gridH / 2 - 7, "PAUSED", COLOR_YELLOW);
        }

        if (gameOver) {
            game.FillRect(gridX + gridW / 2 - 80, gridY + gridH / 2 - 30, 160, 70, COLOR_DARK_GRAY);
            game.DrawRect(gridX + gridW / 2 - 80, gridY + gridH / 2 - 30, 160, 70, COLOR_WHITE);
            game.DrawTextScale(gridX + gridW / 2 - 65, gridY + gridH / 2 - 22, "GAME OVER", COLOR_RED, 8, 8);
            game.DrawPrintf(gridX + gridW / 2 - 50, gridY + gridH / 2 + 2, COLOR_WHITE, "Score: %d", score);
            game.DrawText(gridX + gridW / 2 - 55, gridY + gridH / 2 + 20, "R to restart", COLOR_YELLOW);
        }

        game.Update();
        game.WaitFrame(60);
    }
    return 0;
}

File: examples/09_snake.py
"""09_snake.py - Snake Game

Classic Snake: use arrow keys to control the snake, eat food to grow,
game over if you hit the wall or yourself.
Learn: draw_grid, fill_cell, is_key_pressed, game state machine, timed movement,
       load_sprite, draw_sprite_scaled, play_wav
"""
import os

import pyezgame as g

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))

GRID_ROWS = 20
GRID_COLS = 20
CELL_SIZE = 22
MAX_SNAKE = 400




def main() -> None:
    game = g.GameLib()

    grid_w = GRID_COLS * CELL_SIZE
    grid_h = GRID_ROWS * CELL_SIZE
    win_w = grid_w + 160
    win_h = grid_h + 40
    game.open(win_w, win_h, "09 - Snake", True)

    # Load sprite assets
    food_path = g.get_respath("../clib/assets/fruit_apple.png")
    spr_food = game.load_sprite(food_path)

    # Load sound assets
    eat_sfx = g.get_respath("../clib/assets/sound/coin.wav")
    game_over_sfx = g.get_respath("../clib/assets/sound/game_over.wav")

    grid_x, grid_y = 10, 30

    # Snake body: list of (row, col) tuples
    snake = [(10, 10), (10, 9), (10, 8)]
    snake_len = 3

    # Direction: 0=up 1=down 2=left 3=right
    direction = 3
    next_dir = 3

    # Food
    food_r, food_c = 5, 15

    score = 0
    game_over = False
    paused = False

    move_timer = 0.0
    move_interval = 0.15

    while not game.is_closed():
        if game.is_key_pressed(g.KEY_ESCAPE):
            break

        dt = game.get_delta_time()

        if not game_over:
            if game.is_key_pressed(g.KEY_P):
                paused = not paused

            if not paused:
                if game.is_key_pressed(g.KEY_UP) and direction != 1:
                    next_dir = 0
                if game.is_key_pressed(g.KEY_DOWN) and direction != 0:
                    next_dir = 1
                if game.is_key_pressed(g.KEY_LEFT) and direction != 3:
                    next_dir = 2
                if game.is_key_pressed(g.KEY_RIGHT) and direction != 2:
                    next_dir = 3

                move_timer += dt
                if move_timer >= move_interval:
                    move_timer = 0
                    direction = next_dir

                    # Calculate new head position
                    head_r, head_c = snake[0]
                    if direction == 0:
                        head_r -= 1
                    elif direction == 1:
                        head_r += 1
                    elif direction == 2:
                        head_c -= 1
                    elif direction == 3:
                        head_c += 1

                    # Wall collision
                    if head_r < 0 or head_r >= GRID_ROWS or head_c < 0 or head_c >= GRID_COLS:
                        game_over = True
                        game.play_wav(game_over_sfx, 1, 1000)
                    else:
                        # Self collision
                        if (head_r, head_c) in snake[:snake_len]:
                            game_over = True
                            game.play_wav(game_over_sfx, 1, 1000)

                    if not game_over:
                        ate = (head_r == food_r and head_c == food_c)

                        if not ate:
                            # Move: shift body, drop tail
                            snake.insert(0, (head_r, head_c))
                            snake.pop()
                        else:
                            # Grow
                            snake.insert(0, (head_r, head_c))
                            snake_len += 1
                            score += 10
                            if move_interval > 0.06:
                                move_interval -= 0.003

                            # New food (ensure not on snake)
                            while True:
                                food_r = g.GameLib.random(0, GRID_ROWS - 1)
                                food_c = g.GameLib.random(0, GRID_COLS - 1)
                                if (food_r, food_c) not in snake[:snake_len]:
                                    break
                            game.play_wav(eat_sfx, 1, 800)
        else:
            if game.is_key_pressed(g.KEY_R):
                snake = [(10, 10), (10, 9), (10, 8)]
                snake_len = 3
                direction = 3
                next_dir = 3
                food_r, food_c = 5, 15
                score = 0
                move_interval = 0.15
                game_over = False

        # --- Drawing ---
        game.clear(g.COLOR_BLACK)

        game.draw_text_scale(grid_x, 5, "SNAKE", g.COLOR_GREEN, 16, 16)
        game.draw_grid(grid_x, grid_y, GRID_ROWS, GRID_COLS, CELL_SIZE, g.COLOR_DARK_GRAY)

        # Draw food (use sprite if loaded)
        if spr_food >= 0:
            fx = grid_x + food_c * CELL_SIZE + 1
            fy = grid_y + food_r * CELL_SIZE + 1
            game.draw_sprite_scaled(
                spr_food, fx, fy, CELL_SIZE - 2, CELL_SIZE - 2, g.SPRITE_COLORKEY,
            )
        else:
            game.fill_cell(grid_x, grid_y, food_r, food_c, CELL_SIZE, g.COLOR_RED)

        for i in range(min(snake_len, len(snake))):
            sr, sc = snake[i]
            c = g.COLOR_GREEN if i == 0 else g.COLOR_DARK_GREEN
            game.fill_cell(grid_x, grid_y, sr, sc, CELL_SIZE, c)

        # Right info panel
        info_x = grid_x + grid_w + 15
        game.draw_text(info_x, 40, "Score:", g.COLOR_WHITE)
        game.draw_number(info_x, 55, score, g.COLOR_GOLD)

        game.draw_text(info_x, 85, "Length:", g.COLOR_WHITE)
        game.draw_number(info_x, 100, snake_len, g.COLOR_CYAN)

        game.draw_text(info_x, 140, "Controls:", g.COLOR_GRAY)
        game.draw_text(info_x, 158, "Arrows", g.COLOR_LIGHT_GRAY)
        game.draw_text(info_x, 172, "P: Pause", g.COLOR_LIGHT_GRAY)

        if paused and not game_over:
            game.fill_rect(grid_x + grid_w // 2 - 50, grid_y + grid_h // 2 - 15,
                           100, 30, g.COLOR_DARK_GRAY)
            game.draw_text(grid_x + grid_w // 2 - 30, grid_y + grid_h // 2 - 7,
                           "PAUSED", g.COLOR_YELLOW)

        if game_over:
            cx = grid_x + grid_w // 2
            cy = grid_y + grid_h // 2
            game.fill_rect(cx - 80, cy - 30, 160, 70, g.COLOR_DARK_GRAY)
            game.draw_rect(cx - 80, cy - 30, 160, 70, g.COLOR_WHITE)
            game.draw_text_scale(cx - 65, cy - 22, "GAME OVER", g.COLOR_RED, 8, 8)
            game.draw_printf(cx - 50, cy + 2, g.COLOR_WHITE, f"Score: {score}")
            game.draw_text(cx - 55, cy + 20, "R to restart", g.COLOR_YELLOW)

        game.update()
        game.wait_frame(60)

    if spr_food >= 0:
        game.free_sprite(spr_food)


if __name__ == "__main__":
    main()
